File: test-suite/src/test/java/io/micronaut/security/docs/customauthentication/AuthenticationWithEmail.java
package io.micronaut.security.docs.customauthentication;

import io.micronaut.security.authentication.Authentication;
import io.micronaut.serde.annotation.Serdeable;

@Serdeable
public record AuthenticationWithEmail(String username,
                                      String email) {
    public static AuthenticationWithEmail of(Authentication authentication) {
        Object obj = authentication.getAttributes().get("email");
        String email = obj == null ? null : obj.toString();
        return new AuthenticationWithEmail(authentication.getName(), email);
    }
}

File: test-suite-kotlin/src/test/kotlin/io/micronaut/security/docs/customauthentication/AuthenticationWithEmail.kt
package io.micronaut.security.docs.customauthentication

import io.micronaut.security.authentication.Authentication
import io.micronaut.serde.annotation.Serdeable

@Serdeable
data class AuthenticationWithEmail(val username: String, val email: String?) {
    companion object {
        fun of(authentication: Authentication): AuthenticationWithEmail {
            val obj = authentication.getAttributes()["email"]
            val email = obj?.toString()
            return AuthenticationWithEmail(authentication.name, email)
        }
    }
}
